File: src/prototypes/entity/note_player.lua
local sounds = require('sounds')
local util = require('util')

local emptyWires = { wire = {}, shadow = {}}

local baseEntity = data.raw['power-switch']['power-switch']
function makeNotePlayer(instrument, note)
	return {
		type = 'power-switch',
		name = string.format('musical-speaker-note-player-%s-%s', instrument, note.name),
		icon = '__base__/graphics/icons/programmable-speaker.png',
		icon_size = 64, icon_mipmaps = 4,
		flags = { 'not-on-map', 'hidden', 'not-flammable', 'no-copy-paste', 'not-deconstructable'  },
		collision_bask = {},
		working_sound = {
			sound = {
				filename = note.filename,
				preload = false,
				audible_distance_modifier = 1e99
			},
		},

		wire_max_distance = 5,

		circuit_wire_connection_point = emptyWires,
		left_wire_connection_point = emptyWires,
		right_wire_connection_point = emptyWires,
		led_off = util.empty_sprite(),
		led_on = util.empty_sprite(),
		overlay_loop = util.empty_sprite(),
		overlay_start = util.empty_sprite(),
		overlay_start_delay = 0,
		power_on_animation = util.empty_sprite()
	}
end

entities = {}
for _, category in pairs(sounds) do
	for _, instrument in pairs(category.instruments) do
		for _, note in pairs(instrument.notes) do
			table.insert(entities, makeNotePlayer(instrument.name, note))
		end
	end
end

data:extend(entities)

File: src/prototypes/entity/note-player.ts
import { programmableSpeakerInstruments } from '../../script/sounds';
import * as util from 'util';

const emptyWires: WireConnectionPoint = {
	wire: {},
	shadow: {}
};

const baseSpeaker = data.raw['programmable-speaker']['programmable-speaker'] as PrototypeProgrammableSpeaker;

const debug = false;

let notePlayer = table.deepcopy(baseSpeaker);

notePlayer.name = 'musical-speaker-note-player';
notePlayer.minable = undefined;
notePlayer.flags = [
	'not-on-map', 'hidden', 'not-flammable', 'not-deconstructable', 'player-creation',
	"no-copy-paste", "not-flammable", "not-upgradable", "not-selectable-in-game", "placeable-off-grid"
];
notePlayer.collision_mask = [];
notePlayer.collision_box = undefined;
notePlayer.energy_source = {
	type: 'void'
};

notePlayer.instruments = programmableSpeakerInstruments as any;

if (!debug) {
	notePlayer = {
		...notePlayer,
		...{
			circuit_wire_connection_point: emptyWires,
			sprite: util.empty_sprite(),
			circuit_connector_sprites: {
				...baseSpeaker.circuit_connector_sprites!,
			},
			draw_circuit_wires: false,
			draw_copper_wires: false,
			selectable_in_game: false,
			selection_priority: undefined
		}
	};
}

data.extend([notePlayer]);
